src/MLJ.jl

```diff
@@ -12,25 +17,6 @@ function okay_to_continue(model, fitresult, cache)
          all(_get_struct_loss(model) .== _get_struct_loss(fitresult))
 end
 
-# Generate names to be used by feature_importances in the report
-MMI.reformat(::EvoTypes, X, y, w) =
-  ((matrix=MMI.matrix(X), names=[name for name ∈ schema(X).names]), y, w)
-MMI.reformat(::EvoTypes, X, y) =
-  ((matrix=MMI.matrix(X), names=[name for name ∈ schema(X).names]), y)
-MMI.reformat(::EvoTypes, X) =
-  ((matrix=MMI.matrix(X), names=[name for name ∈ schema(X).names]),)
-MMI.reformat(::EvoTypes, X::AbstractMatrix, y, w) =
-  ((matrix=X, names=["feat_$i" for i = 1:size(X, 2)]), y, w)
-MMI.reformat(::EvoTypes, X::AbstractMatrix, y) =
-  ((matrix=X, names=["feat_$i" for i = 1:size(X, 2)]), y)
-MMI.reformat(::EvoTypes, X::AbstractMatrix) =
-  ((matrix=X, names=["feat_$i" for i = 1:size(X, 2)]),)
-MMI.selectrows(::EvoTypes, I, A, y, w) =
-  ((matrix=view(A.matrix, I, :), names=A.names), view(y, I), view(w, I))
-MMI.selectrows(::EvoTypes, I, A, y) =
-  ((matrix=view(A.matrix, I, :), names=A.names), view(y, I))
-MMI.selectrows(::EvoTypes, I, A) = ((matrix=view(A.matrix, I, :), names=A.names),)
-
```

Comment:
My understanding is this: previously table -> matrix conversion happened in `reformat` and not in EvoTree's internal training, which means MLJ's `IteratedModel` wrapper avoided repeating these conversions over and over in controlled iteration. 

The change here suggests that table -> matrix conversion is now "internalised", which means the performance gain is lost, right? Or am I missing something.

Reply:
I'm afraid this brings back at the forefront the issue with regard to MLJ's handling of iterated process for models relying on caching / initialization mechanism. 

Once an EvoTree model is initialized with a training dataset, this dataset should be considered as fixed, no subsampling should be performed to it throughout the iterations. The algo is responsible for performing the subsampling operations throughout the iterations using tricks that are algo specific. This is true AFAIK for all "high performance" GBT models (incl. XGBoost, LightGBM, CatBoost). 

Looking at the error resulting when trying an IteratedModel uder this PR, why MLJ's `fit_only!` needs to perform a data resampling:  `fit(model, verbosity, _resampled_data(mach, model, rows)...)`?

The logic that has been used with EvoTrees so far to handle the MMI's `update` method has been to completely ignore the data args `A` and `y`: https://github.com/Evovest/EvoTrees.jl/blob/20b9e2a4a14863e87e16fe70fbece98c02b46ba2/src/MLJ.jl#L34
I'm still unclear why an issue happened with `IteratedModel` at first since `update` behavior where only the `model` and `cache` are actually used when iterating the model. And the EvoTrees' MLJ tests pass, so the basic operations seem to work fine, including iterative calls to `fit!`. 

This is still much WIP, but I'm trying to lay down what I have experiencced as shortcomings from ML toolkits, incl Scikitlearn, which assumes that a model can be instantiated with a `MyModelRegressor` rather than having distinct management of a model conguration, its initialization and the fitting of that instantiated model: https://github.com/Evovest/EvoTrees.jl/blob/learnAPI/experiments/learnAPI.jl. This is the kind of approach that we need to rely on internally to handle our ML pipelines, but I'd definitely like to see such support in an open framework.

Reply:
@ablaom The issue I encountered testing on `IteratedModel` appeared to be a red herring: the failure was caused by the `log_loss_` measure not accepting `Deterministic` output of the EvoTreeRegressor (isn't it unnecessarily restrictive? I actually don't see a logistic regression as less deterministic than "mse", but that's another topic!).  
It seems like the MLJ's `fit!` call to to reformat effectively results in a no-op as expected (each iteration following the 1st one directly mutate the instantiate fitted_model and the cache). Training speed is aligned with internal's `fit_evotree` if using very large evaluation `Step` (but quickly blows up, going from 12 sec reference time to ~68 sec with `Step(5)`). 

So back the this specific topic of moving the EvoTrees's table API, AFAICT the MLJ integration with MLJ works fine. Or at least as well as the existing one, but with added benefits in terms of richer input feature types.